BrowsingContextGroup should only be shared from opener to opened WKWebViews

achristensen07 opened the pull request:
#### ca2fc5fb29dc49a4b9696ecbeb4aeb175c4234d3
<pre>
BrowsingContextGroup should only be shared from opener to opened WKWebViews
<a href="https://bugs.webkit.org/show_bug.cgi?id=293027">https://bugs.webkit.org/show_bug.cgi?id=293027</a>
<a href="https://rdar.apple.com/151345757">rdar://151345757</a>

Reviewed by Sihui Liu.

When WKUIDelegate.createWebViewWithConfiguration is called, it gives the application
a WKWebViewConfiguration to use to make a WKWebView and put it in the same
BrowsingContextGroup.  If a WKWebViewConfiguration is reused, we need to not
use the same BrowsingContextGroup, which is web compatible because there is no
opener relationship to maintain in that case.  To implement this, we consume
the configuration&apos;s opener info when making a WebPageProxy so the configuration
can be further copied and reused without adding more pages to the BrowsingContextGroup.

An app can copy the opener&apos;s configuration and use that configuration to load
the opener site in an auxiliary WKWebView before responding to
WKUIDelegate.createWebViewWithConfiguration. When this happens with site isolation off,
the auxiliary WKWebView loads in its own process.  When this happens with site isolation
on before this PR, it would try to load in the process that was waiting for the reply
to Messages::WebPageProxy::CreateNewPage, which caused a deadlock.  This solution
avoids the deadlock, increases compatibility, and allows the authentication to proceed.

This fixes the last API test in the SOAuthorizationPopUp group with site isolation on.

* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::Data::Data):
(API::PageConfiguration::openerInfo const const):
(API::PageConfiguration::setOpenerInfo):
(API::PageConfiguration::consumeOpenerInfo):
(API::PageConfiguration::Data::createBrowsingContextGroup): Deleted.
(API::PageConfiguration::browsingContextGroup const): Deleted.
(API::PageConfiguration::setBrowsingContextGroup): Deleted.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::Data::initializer):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::createNewPage):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::m_pageForTesting):
(WebKit::WebPageProxy::initializeWebPage):
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, LoadDuringOpen)):
(TestWebKitAPI::TEST(SiteIsolation, ChildNavigatingToDomainLoadedOnADifferentPage)):
(TestWebKitAPI::TEST(SiteIsolation, ReuseConfigurationLoadHTMLString)):

Canonical link: <a href="https://commits.webkit.org/294968@main">https://commits.webkit.org/294968@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c7726c67de87079e961d124d8b04d3d751bd859e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103606 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23288 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13609 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108782 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54253 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105646 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23644 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31839 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78728 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106612 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18340 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93464 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59063 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18152 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11511 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53607 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87933 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11570 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111160 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30753 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22677 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87724 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31114 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89664 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87371 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32246 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9968 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25107 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16824 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30681 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35993 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30481 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33812 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32042 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -639,6 +639,28 @@ HTTPServer server({
+    }, HTTPServer::Protocol::HttpsProxy);
+    auto [webView, navigationDelegate] = siteIsolatedViewAndDelegate(server);
+    RetainPtr uiDelegate = adoptNS([TestUIDelegate new]);
+    uiDelegate.get().createWebViewWithConfiguration = [&](WKWebViewConfiguration *configuration, WKNavigationAction *action, WKWindowFeatures *windowFeatures) -> WKWebView * {
+        RetainPtr auxiliary = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectZero configuration:[webView configuration]]);
+        auxiliary.get().navigationDelegate = navigationDelegate.get();
+        [auxiliary loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"https://example.com/example"]]];
+        [navigationDelegate waitForDidFinishNavigation];
+        return nil;
+    };
```
My understanding with this approach is:
- view1 loads example.com and requests a new view
- `createWebViewWithConfiguration` is invoked with configuration that has opener iInfo
- Client creates view2 with the configuration, which clears opener info in configuration and set the opener info on view2 (`WebPageProxy::m_browsingContextGroup`)
- view2 loads example.com and it will be loaded in a different process than view1
- `createWebViewWithConfiguration` returns

My question is what makes view2 load in a different process when it gets the opener info from the configuration? (Maybe I am missing something)

achristensen07 replied:
If view 2 is created from the provided configuration in createWebViewWithConfiguration then it will share the same BrowsingContextGroup with view 1 and will load example.com in the same process as view 1 has loaded example.com in.  However, in this test auxiliary is created with a copy of view 1's configuration, not the provided configuration, so it needs to have a different BrowsingContextGroup.

szewai replied:
I see! Do we need to support the case where client creates view2 with the provided configuration, and loads example.com? (Will this work without site isolation?)
Based on your investigation, maybe we should set a flag on `WebProcessProxy` when it's blocked waiting on a sync message reply and check the flag when the `WebProcessProxy` is picked for loading.

achristensen07 replied:
No, that never worked even without site isolation.

achristensen07 committed:
BrowsingContextGroup should only be shared from opener to opened WKWebViews
https://bugs.webkit.org/show_bug.cgi?id=293027
rdar://151345757

Reviewed by Sihui Liu.

When WKUIDelegate.createWebViewWithConfiguration is called, it gives the application
a WKWebViewConfiguration to use to make a WKWebView and put it in the same
BrowsingContextGroup.  If a WKWebViewConfiguration is reused, we need to not
use the same BrowsingContextGroup, which is web compatible because there is no
opener relationship to maintain in that case.  To implement this, we consume
the configuration's opener info when making a WebPageProxy so the configuration
can be further copied and reused without adding more pages to the BrowsingContextGroup.

An app can copy the opener's configuration and use that configuration to load
the opener site in an auxiliary WKWebView before responding to
WKUIDelegate.createWebViewWithConfiguration. When this happens with site isolation off,
the auxiliary WKWebView loads in its own process.  When this happens with site isolation
on before this PR, it would try to load in the process that was waiting for the reply
to Messages::WebPageProxy::CreateNewPage, which caused a deadlock.  This solution
avoids the deadlock, increases compatibility, and allows the authentication to proceed.

This fixes the last API test in the SOAuthorizationPopUp group with site isolation on.

* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::Data::Data):
(API::PageConfiguration::openerInfo const const):
(API::PageConfiguration::setOpenerInfo):
(API::PageConfiguration::consumeOpenerInfo):
(API::PageConfiguration::Data::createBrowsingContextGroup): Deleted.
(API::PageConfiguration::browsingContextGroup const): Deleted.
(API::PageConfiguration::setBrowsingContextGroup): Deleted.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::Data::initializer):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::createNewPage):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::m_pageForTesting):
(WebKit::WebPageProxy::initializeWebPage):
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, LoadDuringOpen)):
(TestWebKitAPI::TEST(SiteIsolation, ChildNavigatingToDomainLoadedOnADifferentPage)):
(TestWebKitAPI::TEST(SiteIsolation, ReuseConfigurationLoadHTMLString)):

Canonical link: https://commits.webkit.org/294968@main
